tensorboard/loader.py

```diff
@@ -0,0 +1,413 @@
+@six.python_2_unicode_compatible
+class BufferedRecordReader(object):
+  """Wrapper around RecordReader that does threaded read-ahead.
+
+  This class implements the same interface as RecordReader. It prevents
+  remote file systems from devastating loader performance. It does not
+  degrade throughput on local file systems.
+
+  The thread is spawned when the first read operation happens. The
+  thread will diligently try to buffer records in the background. Its
+  goal is to sleep as much as possible without blocking read operations.
+
+  This class is thread safe. It can be used from multiple threads
+  without any need for external synchronization.
+  """
+
+  READ_AHEAD_AGGRESSION = 2.3  # Does full replenish when ~40% full.
+  READ_AHEAD_BYTES = 16 * 1024 * 1024
+  STAT_INTERVAL_SECONDS = 4.0
+
+  def __init__(self, path,
+               start_offset=0,
+               read_ahead=None,
+               stat_interval=None,
+               clock=time.time,
+               record_reader_factory=RecordReader):
+    """Creates new instance.
+
+    The i/o thread is not started until the first read happens.
+
+    Args:
+      path: Path of file. This can be on a remote file system if the
+          TensorFlow build supports it.
+      start_offset: Byte offset to seek in file once it's opened.
+      read_ahead: The number of record bytes to buffer into memory
+          before the thread starts blocking. This value must be >0 and
+          the default is BufferedRecordReader.READ_AHEAD_BYTES.
+      stat_interval: A float with the minimum number of seconds between
+          stat calls, to determine the file size.
+      clock: Function returning a float with the number of seconds
+          since the UNIX epoch in zulu time.
+      record_reader_factory: The RecordReader constructor, which can be
+          changed for testing.
+
+    :type path: str
+    :type start_offset: int
+    :type read_ahead: int
+    :type clock: () -> float
+    """
+    self.path = tf.compat.as_text(path)
+    self._read_ahead = read_ahead or BufferedRecordReader.READ_AHEAD_BYTES
+    self._stat_interval = (stat_interval
+                           if stat_interval is not None else
+                           BufferedRecordReader.STAT_INTERVAL_SECONDS)
+    self._clock = clock
+    self._is_closed = False
+    self._has_reached_end = False
+    self._offset = 0
+    self._size = -1
+    self._last_stat = 0.0
+    self._buffered = 0
+    self._reader = record_reader_factory(self.path, start_offset)
+    self._records = collections.deque()  # type: collections.deque[Record]
+    self._read_exception = None  # type: Exception
+    self._close_exception = None  # type: Exception
+    self._lock = threading.Lock()
+    self._wake_up_producer = threading.Condition(self._lock)
+    self._wake_up_consumers = threading.Condition(self._lock)
+    self._thread = threading.Thread(target=self._run,
+                                    name=_shorten_event_log_path(self.path))
```

Comment:
It might be better to use `"BufferedRecordReaderWorker_" + _shorten…` (or similar); having a thread named `mnist/train` might be clear, but `.` is not.